File: src/shaman_memory.erl
%% Copyright (c) 2014, Loïc Hoguin <essen@ninenines.eu>
%%
%% Permission to use, copy, modify, and/or distribute this software for any
%% purpose with or without fee is hereby granted, provided that the above
%% copyright notice and this permission notice appear in all copies.
%%
%% THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
%% WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
%% MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
%% ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
%% WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
%% ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
%% OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.

-module(shaman_memory).
-behaviour(alien_process_probe).
-behaviour(gen_server).

-export([start_link/2]).
-export([init/1]).
-export([handle_call/3]).
-export([handle_cast/2]).
-export([handle_info/2]).
-export([terminate/2]).
-export([code_change/3]).

-record(state, {
	ref :: alien:ref(),
	interval :: pos_integer()
}).

start_link(Ref, []) ->
	gen_server:start_link(?MODULE, [Ref], []).

init([Ref]) ->
	Interval = 1000, %% @todo Config.
	_ = erlang:send_after(Interval, self(), update),
	{ok, #state{ref=Ref, interval=Interval}}.

handle_call(_Request, _From, State) ->
	{reply, ignored, State}.

handle_cast(_Msg, State) ->
	{noreply, State}.

handle_info(update, State=#state{ref=Ref, interval=Interval}) ->
	alien:event(Ref,
		[{atom_to_binary(K, utf8), V} || {K, V} <- erlang:memory([
			total, processes, atom, binary, code, ets])]),
	_ = erlang:send_after(Interval, self(), update),
	{noreply, State};
handle_info(_Info, State) ->
	{noreply, State}.

terminate(_Reason, _State) ->
	ok.

code_change(_OldVsn, State, _Extra) ->
	{ok, State}.

File: priv/static/shaman-memory.js
// Copyright (c) 2014, Loïc Hoguin <essen@ninenines.eu>
//
// Permission to use, copy, modify, and/or distribute this software for any
// purpose with or without fee is hereby granted, provided that the above
// copyright notice and this permission notice appear in all copies.
//
// THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
// WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
// MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
// ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
// WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
// ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
// OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.

var shaman_memory = (function(){
	var self = {};

	self.link = "#memory";
	self.title = "Memory";
	self.html = '\
<div id="memory" class="container shaman-page">\
</div>';

	var memory = {};

	self.onInit = function(){};
	self.onNodeChange = function(){};

	self.onData = function(node, data){
		if (!memory[node]){
			$("#memory").append('<svg data-node="' + node + '"></svg>');

			memory[node] = {};
			memory[node].d = [];
			memory[node].g = new shaman.d3.stacked_area_chart(
				'#memory svg[data-node="' + node + '"]', {
					names: ["code", "atom", "ets", "binary", "processes"],
					max: function(d){return shaman.util.arrayKeyMax(d, "total")}
				});
		}

		memory[node].d.push(data);
		memory[node].d = memory[node].d.slice(-60);
		memory[node].g.update(memory[node].d);
	}

	return self;
}());

$(document).ready(function(){
	shaman
		.add_page(shaman_memory)
		.add_data_handler("memory", shaman_memory);
});
